libs/pilight/events/actions/sendmail.c

```diff
@@ -0,0 +1,238 @@
+		return -1;
+	}
+	if(settings_find_string("smtp-password", &stmp) != EXIT_SUCCESS) {
+		logprintf(LOG_ERR, "Sendmail: setting \"smtp-password\" is missing in config");
+		return -1;
+	}
+	if(settings_find_string("smtp-sender", &stmp) != EXIT_SUCCESS) {
+		logprintf(LOG_ERR, "Sendmail: setting \"smtp-sender\" is missing in config");
+		return -1;
+	}
+	return 0;
+}
+
+static void *actionSendmailThread(void *param) {
+	struct JsonNode *arguments = (struct JsonNode *)param;
+	struct JsonNode *jsubject = NULL;
+	struct JsonNode *jmessage = NULL;
+	struct JsonNode *jto = NULL;
+	struct JsonNode *jvalues1 = NULL;
+	struct JsonNode *jvalues2 = NULL;
+	struct JsonNode *jvalues3 = NULL;
+	struct JsonNode *jval1 = NULL;
+	struct JsonNode *jval2 = NULL;
+	struct JsonNode *jval3 = NULL;
+
+	action_sendmail->nrthreads++;
+
+	struct mail_t mail;
+	char *shost = NULL, *suser = NULL, *spassword = NULL;
+	int sport= 0;
+
+	jmessage = json_find_member(arguments, "MESSAGE");
+	jsubject = json_find_member(arguments, "SUBJECT");
+	jto = json_find_member(arguments, "TO");
+
+	if(jsubject != NULL && jmessage != NULL && jto != NULL) {
+		jvalues1 = json_find_member(jsubject, "value");
+		jvalues2 = json_find_member(jmessage, "value");
+		jvalues3 = json_find_member(jto, "value");
+		if(jvalues1 != NULL && jvalues2 != NULL && jvalues3 != NULL) {
+			jval1 = json_find_element(jvalues1, 0);
+			jval2 = json_find_element(jvalues2, 0);
+			jval3 = json_find_element(jvalues3, 0);
+			if(jval1 != NULL && jval2 != NULL && jval3 != NULL &&
+				jval1->tag == JSON_STRING && jval2->tag == JSON_STRING && jval3->tag == JSON_STRING) {
+
+				//smtp settings
+				settings_find_string("smtp-sender", &mail.from);
+				settings_find_string("smtp-host", &shost);
+				settings_find_number("smtp-port", &sport);
+				settings_find_string("smtp-user", &suser);
+				settings_find_string("smtp-password", &spassword);
+
+				mail.subject = jval1->string_;
+				mail.message = jval2->string_;
+				mail.to = jval3->string_;
+
+				if(sendmail(shost, suser, spassword, sport, &mail) != 0) {
+					logprintf(LOG_ERR, "Sendmail failed to send message \"%s\"", jval2->string_);
+				}
+			}
+		}
+	}
+
+	action_sendmail->nrthreads--;
+	return (void *)NULL;
+}
+
+static int actionSendmailRun(struct JsonNode *arguments) {
+	pthread_t pth;
+	threads_create(&pth, NULL, actionSendmailThread, (void *)arguments);
+	pthread_detach(pth);
+	return 0;
+}
+
+#if !defined(MODULE) && !defined(_WIN32)
+__attribute__((weak))
+#endif
+void actionSendmailInit(void) {
+	event_action_register(&action_sendmail, "sendmail");
+
+	options_add(&action_sendmail->options, 'a', "SUBJECT", OPTION_HAS_VALUE, DEVICES_VALUE, JSON_STRING, NULL, NULL);
+	options_add(&action_sendmail->options, 'b', "MESSAGE", OPTION_HAS_VALUE, DEVICES_VALUE, JSON_STRING, NULL, NULL);
+	options_add(&action_sendmail->options, 'c', "TO", OPTION_HAS_VALUE, DEVICES_VALUE, JSON_STRING, NULL, NULL);
```

Comment:
No, indeed it didn't work here, so I did it in the arguments check.

CurlyMoo schreef op 2-4-2015 om 21:17:
>
> In libs/pilight/events/actions/sendmail.c 
> <https://github.com/pilight/pilight/pull/166#discussion_r27688082>:
>
> > +static int actionSendmailRun(struct JsonNode *arguments) {
> > +	pthread_t pth;
> > +	threads_create(&pth, NULL, actionSendmailThread, (void *)arguments);
> > +	pthread_detach(pth);
> > +	return 0;
> > +}
> > +
> > +#if !defined(MODULE) && !defined(_WIN32)
> > +__attribute__((weak))
> > +#endif
> > +void actionSendmailInit(void) {
> > +	event_action_register(&action_sendmail, "sendmail");
> > +
> > +	options_add(&action_sendmail->options, 'a', "SUBJECT", OPTION_HAS_VALUE, DEVICES_VALUE, JSON_STRING, NULL, NULL);
> > +	options_add(&action_sendmail->options, 'b', "MESSAGE", OPTION_HAS_VALUE, DEVICES_VALUE, JSON_STRING, NULL, NULL);
> > +	options_add(&action_sendmail->options, 'c', "TO", OPTION_HAS_VALUE, DEVICES_VALUE, JSON_STRING, NULL, NULL);
>
> Didn't the mail regex work here?
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/pilight/pilight/pull/166/files#r27688082>.
>